### PR TITLE
Add chemfiles and py-chemfiles

### PR DESCRIPTION
**Pull request by RMeli**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyChemfiles(CMakePackage):
+    """Python interface to chemfiles"""
+
+    homepage = "http://chemfiles.org/chemfiles.py/latest/"
+    url = "https://github.com/chemfiles/chemfiles.py/archive/refs/tags/0.10.3.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    maintainers("RMeli")
+
+    version("0.10.3", sha256="0e12837a332e6c1c950e82e339b17cd8fbf29a28bae8399a664b818c056786a9")
+    version("0.10.2", sha256="a94e7fb9bae7f0b1658b9b3daab576cf98ea86779ad812abe9e70f49ee0cec48")
+    version("0.10.1", sha256="32cee8caa9a626340e3a4b74aa7e50559cf06fae80d052dbb8bd223b47cbee63")
+    version("0.10.0", sha256="b4438692d69a0e325157ff0b0ca441ae9b790a2b3104d92cd5b30920cd07fabe")
+
+    extends("python")
+    depends_on("chemfiles+shared")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]: Most deps are missing, see pyproject.toml and setup.py

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-chemfiles/package.py (hunk 2 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/py-chemfiles/package.py]: Thanks @adamjstewart. It's currently a draft PR, I'm still working on the Python recipe.

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    maintainers("RMeli")
+
+    version("0.10.3", sha256="4bbb8b116492a57dbf6ddb4c84aad0133cd782e0cc0e53e4b957f2d93e6806ea")
+    version("0.10.2", sha256="e277725803715762f9ea787f1ed51dbc2a83a47188ca3bf5d77ddcbc527f55f9")
+    version("0.10.1", sha256="6fe35c529c2ded099a59b689270ac0368c6aa33664069c1ccf88eb9fb2686906")
+    version("0.10.0", sha256="b52bc19ac7967935a2acc896b1b9738bd904444a1fc09589ed6ada7658a02bf4")
+
+    depends_on("python")
+    depends_on("chemfiles+shared")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]: Do the versions of chemfiles and py-chemfiles need to match?

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    depends_on("py-numpy")
+
+    depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]:
```suggestion
    depends_on("py-setuptools@44:", type="build")
```

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    depends_on("py-wheel", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]:
```suggestion
    depends_on("py-wheel@0.36:", type="build")
```

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    maintainers("RMeli")
+
+    version("0.10.3", sha256="4bbb8b116492a57dbf6ddb4c84aad0133cd782e0cc0e53e4b957f2d93e6806ea")
+    version("0.10.2", sha256="e277725803715762f9ea787f1ed51dbc2a83a47188ca3bf5d77ddcbc527f55f9")
+    version("0.10.1", sha256="6fe35c529c2ded099a59b689270ac0368c6aa33664069c1ccf88eb9fb2686906")
+    version("0.10.0", sha256="b52bc19ac7967935a2acc896b1b9738bd904444a1fc09589ed6ada7658a02bf4")
+
+    depends_on("python")
+    depends_on("chemfiles+shared")
+    depends_on("py-numpy")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]:
```suggestion
    depends_on("py-numpy", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-cmake/package.py [hunk 7]
```diff
@@ -15,6 +15,15 @@ class PyCmake(PythonPackage):
...
+    version("3.24.3", sha256="73fa252b21fce1db988beace3e1a2b21f926dd870322d9f37ff04e773b13a12a")
+    version("3.24.2", sha256="c62f317b9c9efb8ded26b910b966569ee6df335f21c8017617cd45d985b985c9")
+    version("3.24.1", sha256="204a5e26a1e609c2e8e31db3fe16bb016d358fb31d457e08a8525267afe4eb88")
+    version("3.24.0", sha256="4b16db74e79a5eb8e2b1a513157a94a20344987ab41acb2e0676fce9baf9c1aa")
+    version("3.23.3", sha256="72dbd32dbabf586b3f8148d8cd807a942bdcf0eb5d0178f5826d2221a947aa96")
+    version("3.22.6", sha256="4731681ceb0f001d109cf322acb1f3b9896bceb76882979687552972502ec43f")
+    version("3.22.5", sha256="4362d75d12d07bdaf1a467e3edefd866a99d97704bbc2464eb5b6c16695c20b9")
     version("3.22.2", sha256="b5bd5eeb488b13cf64ec963800f3d979eaeb90b4382861b86909df503379e219")
     version("3.21.4", sha256="30fa5ed8a5ad66dcd263adb87f3ce3dc2d0ec0ac3958f5becff577e4b62cd065")
     version("3.18.0", sha256="52b98c5ee70b5fa30a8623e96482227e065292f78794eb085fdf0fecb204b79b")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cmake/package.py]: Need to update dependencies on cmake versions, you can use a for-loop to make it simpler

File: var/spack/repos/builtin/packages/py-ninja/package.py [hunk 8]
```diff
@@ -14,12 +14,17 @@ class PyNinja(PythonPackage):
...
+    version("1.11.1", sha256="c833a47d39b2d1eee3f9ca886fa1581efd5be6068b82734ac229961ee8748f90")
+    version("1.10.2.4", sha256="da7a6d9b2ed2018165fbf90068e2c64da08f2568c700fdb8abea07a245dc4664")
+    version("1.10.2.3", sha256="e1b86ad50d4e681a7dbdff05fc23bb52cb773edb90bc428efba33fa027738408")
+    version("1.10.2.2", sha256="3f8a75acd929abb9f003d3aa5bc299cea30b9db0dfa18669877e9c02ddcf530d")
+    version("1.10.2.1", sha256="719ab357f5dc822711c1151d1b9517c5543340e23f6cc4e8508f793848a48bb1")
     version("1.10.2", sha256="bb5e54b9a7343b3a8fc6532ae2c169af387a45b0d4dd5b72c2803e21658c5791")
 
     depends_on("cmake@3.6:", type="build")
     depends_on("py-setuptools@42:", type="build")
     depends_on("py-scikit-build", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ninja/package.py]: Latest version required skbuild 0.12+

File: var/spack/repos/builtin/packages/py-ninja/package.py [hunk 9]
```diff
@@ -14,12 +14,17 @@ class PyNinja(PythonPackage):
...
-    depends_on("ninja@1.10.2", type=("build", "run"), when="@1.10.2")
+    depends_on("ninja@1.10.2", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ninja/package.py]: Versions of py-ninja must match ninja, you can use a for-loop to simplify this

**Review by adamjstewart — CHANGES_REQUESTED** [review]
The py-cmake and py-ninja changes don't need to be in this PR and should probably done in a separate PR

**Comment by RMeli** [comment]
Thanks @adamjstewart. I simplified the PR a bit, incorporating your suggestions.

> The py-cmake and py-ninja changes don't need to be in this PR

I removed th changes to `py-cmake`, but `py-ninja@1.10.2` failed to build on my end, while updating to a newer version does work as expected.

File: var/spack/repos/builtin/packages/py-chemfiles/package.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    homepage = "http://chemfiles.org/chemfiles.py/latest/"
+    pypi = "chemfiles/chemfiles-0.10.3.tar.gz"
+
+    maintainers("RMeli")
+
+    version("0.10.3", sha256="4bbb8b116492a57dbf6ddb4c84aad0133cd782e0cc0e53e4b957f2d93e6806ea")
+
+    depends_on("chemfiles@0.10.3+shared", when="@0.10.3")
+
+    depends_on("python")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chemfiles/package.py]: This can be removed, it's already added by the base class

**Review by adamjstewart — COMMENTED** [review]
Technically the py-ninja changes could still be moved to a separate PR. The tests for this PR will still pass without it.

File: var/spack/repos/builtin/packages/py-chemfiles/package.py (hunk 10 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/py-chemfiles/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/py-ninja/package.py [hunk 11]
```diff
@@ -14,12 +14,15 @@ class PyNinja(PythonPackage):
     homepage = "https://ninja-build.org"
     pypi = "ninja/ninja-1.10.2.tar.gz"
 
+    version("1.11.1", sha256="c833a47d39b2d1eee3f9ca886fa1581efd5be6068b82734ac229961ee8748f90")
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/py-ninja/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/py-ninja/package.py [hunk 12]
```diff
@@ -14,12 +14,15 @@ class PyNinja(PythonPackage):
...
     version("1.10.2", sha256="bb5e54b9a7343b3a8fc6532ae2c169af387a45b0d4dd5b72c2803e21658c5791")
 
     depends_on("cmake@3.6:", type="build")
     depends_on("py-setuptools@42:", type="build")
     depends_on("py-scikit-build", type="build")
+    depends_on("py-scikit-build@0.12:", type="build", when="@1.11.1")
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/py-ninja/package.py]:
```suggestion

```

File: var/spack/repos/builtin/packages/py-ninja/package.py [hunk 13]
```diff
@@ -14,12 +14,15 @@ class PyNinja(PythonPackage):
...
     depends_on("ninja@1.10.2", type=("build", "run"), when="@1.10.2")
+    depends_on("ninja@1.11.1", type=("build", "run"), when="@1.11.1")
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/py-ninja/package.py]:
```suggestion

```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**